### ISSUE TITLE
Glacier Job payload should be in JSON but it isn't

### ISSUE THREAD

**Issue by adam-fowler** [swift-aws/aws-sdk-swift#99]
Whenever I try to send an InitiateJob() request to Glacier, it always returns the following error.
```
AWSError(message: "SerializationException", rawBody: "{\"code\":\"SerializationException\",\"type\":\"Client\",\"message\":\"SerializationException\"}")
```
The cause of this look like it is because the body of the InitiateJob() request contains the JobParameters and these should be in json. But it is coming out in this form
```
JobParameters(retrievalByteRange: nil, tier: nil, format: nil, selectParameters: nil, description: Optional(\"Inventory\"), inventoryRetrievalParameters: Optional(Glacier.Glacier.InventoryRetrievalJobInput(marker: nil, limit: nil, endDate: nil, startDate: nil)), outputLocation: nil, type: Optional(\"inventory-retrieval\"), archiveId: nil, sNSTopic: nil)
```
This looks like it could be a common issue across all AWS systems where a json body is required.

**Comment by jonnymacs** [comment]
Interesting. There are about 90 services that specify the "json" protocol,
one of which is kinesis which I am using the client to post JSON without
any errors.

However another 55 specify "rest-json", one of which is glacier. These
services may be affected by this issue.

**Comment by adam-fowler** [comment]
The JobParameters are output as the payload. And it is the output of the payload that is the problem. The Kinesis payloads look like they are raw data so wouldn't be affected by this issue.

**Comment by adam-fowler** [comment]
I temporarily sorted it by checking if the payload was an AWSShape and if it was then outputting it as JSON. To get that to work was like pulling teeth, I'm not very experienced with the reflection code so wasted a lot of time. I ending up writing the full Input to json, parsing it into a dictionary and then writing the payload element of the dictionary back out to json.

**Comment by jonnymacs** [comment]
ah, ok I get it now. I'll have to ask @noppoMan if how this ever worked?

Could you try

```
switch payloadBody {
    case is AWSShape:
        body = .json(try AWSShapeEncoder().encodeToJSONUTF8Data(payloadBody as! Input))
    default:
        body = Body(anyValue: payloadBody)
}
```

**Comment by adam-fowler** [comment]
> Could you try
> 
> ```
> switch payloadBody {
>     case is AWSShape:
>         body = .json(try AWSShapeEncoder().encodeToJSONUTF8Data(payloadBody as! Input))
>     default:
>         body = Body(anyValue: payloadBody)
> }
> ```
This doesn't work as the payloadBody is not the same class as the Input shape. I get the following error.
```
Could not cast value of type 'Glacier.Glacier.JobParameters' (0x10058d7f8) to 'Glacier.Glacier.InitiateJobInput' (0x10058d1f0).
```

**Comment by jonnymacs** [comment]
right. ok. Input in this context is the input data type not a generic AWSShape. Meanwhile payloadBody is type Any at compile time... :thinking_face: 

sheesh yeah this is a doozy. I'm down 2 teeth already. Feel free to submit a PR (or share a gist) of what you have working.

**Comment by adam-fowler** [comment]
Here you go
```
                switch payloadBody {
                case is AWSShape:
                    let data = try AWSShapeEncoder().encodeToJSONUTF8Data(input)
                    var params = try JSONSerializer().serializeToDictionary(data)
                    if let payloadDict = params[payload] {
                        let json = try JSONSerialization.data(withJSONObject: payloadDict)
                        body = .json(json)
                    }
                default:
                    body = Body(anyValue: payloadBody)
                }
```

**Comment by jonnymacs** [comment]
Thanks @adam-fowler !

https://github.com/swift-aws/aws-sdk-swift-core/pull/46

This was an issue for the xml apis as well.

**Comment by jonnymacs** [comment]
fixed in #102
